yew-functional-macro/src/lib.rs

```diff
@@ -0,0 +1,78 @@
+use proc_macro::{Span, TokenStream};
+use quote::quote;
+use syn::parse::{Parse, ParseStream};
+use syn::{parse_macro_input, Block, FnArg, Ident, Item, Type};
+
+struct FunctionalComponent {
+    body: Block,
+    props_type: Type,
+}
+
+impl Parse for FunctionalComponent {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let parsed: Item = input.parse()?;
+
+        match parsed {
+            Item::Fn(func) => {
+                let inputs = &func.sig.inputs;
+                let props_type = match inputs.len() {
+                    // If there is one and only one argument passed, use it for props
+                    1 => match inputs.first().unwrap() {
+                        FnArg::Typed(arg) => (*arg.ty).clone(),
+
+                        _ => {
+                            return Err(syn::Error::new_spanned(&func.sig, "invalid argument passed. (hint: first argument must not be a receiver. Functional components take one argument of props struct as value)"));
+                        }
+                    },
+                    // If nothing is passed then use `()` for props
+                    0 => Type::Tuple(syn::TypeTuple {
+                        paren_token: Default::default(),
+                        elems: Default::default(),
+                    }),
+                    // Error if more than 1 arguments are passed
+                    _ => {
+                        return Err(syn::Error::new_spanned(
+                            &func.sig,
+                            "Functional components take only one argument of props struct as value",
+                        ))
+                    }
+                };
+
+                let body = *func.block;
```

Comment:
I take it out so I don't have to `Box` the field on the struct. It helps read better since you don't have to know why `Block` is `Box`ed in the struct field. I think its better as is, lemme know if this should be changed.

Reply:
I agree it's more readable in the code, but I think the performance of moving something potentially quite large from the heap to the stack isn't ideal. Keeping it boxed should (hopefully) help with compile times, even if it's just a tiny bit.

Reply:
Alright, I've `Box`ed it.